internal/cache/file/cache_handle_test.go

```diff
@@ -16,307 +16,419 @@ package file
+	dst := make([]byte, ReadContentSize)
+	offset := int64(5)
+	cht.cacheHandle.fileHandle = nil
 
-	ExpectEq(false, t.ch.IsSequential(int64(currentOffset)))
+	n, err := cht.cacheHandle.Read(context.Background(), cht.object, true, offset, dst)
+	ExpectNe(nil, err)
+	ExpectEq(0, n)
+	ExpectEq(util.InvalidFileHandleErrMsg, err.Error())
 }
 
-func (t *cacheHandleTest) TestIsSequentialWhenOffsetDiffIsMoreThanMaxAllowed() {
-	t.ch.isSequential = true
-	t.ch.prevOffset = 5
-	currentOffset := 8 + downloader.ReadChunkSize
+func (cht *cacheHandleTest) Test_Read_Random() {
+	dst := make([]byte, ReadContentSize)
+	offset := int64(cht.object.Size - ReadContentSize)
+	cht.cacheHandle.isSequential = false
 
-	ExpectEq(false, t.ch.IsSequential(int64(currentOffset)))
+	// Since, it's a random read hence will not wait to download till requested offset.
+	n, err := cht.cacheHandle.Read(context.Background(), cht.object, true, offset, dst)
+
+	jobStatus := cht.cacheHandle.fileDownloadJob.GetStatus()
+	ExpectLt(jobStatus.Offset, offset)
+	ExpectEq(n, 0)
+	ExpectNe(nil, err)
+	ExpectTrue(strings.Contains(err.Error(), util.FallbackToGCSErrMsg))
 }
 
-func (t *cacheHandleTest) TestIsSequentialWhenOffsetDiffIsLessThanMaxAllowed() {
-	t.ch.isSequential = true
-	t.ch.prevOffset = 5
-	currentOffset := 10
+func (cht *cacheHandleTest) Test_Read_RandomWithNoRandomDownload() {
+	dst := make([]byte, ReadContentSize)
+	offset := int64(cht.object.Size - ReadContentSize)
+	cht.cacheHandle.isSequential = false
+	downloadForRandom := false
+
+	// Since, it's a random read hence will not wait to download till requested offset.
+	n, err := cht.cacheHandle.Read(context.Background(), cht.object, downloadForRandom, offset, dst)
 
-	ExpectEq(true, t.ch.IsSequential(int64(currentOffset)))
+	jobStatus := cht.cacheHandle.fileDownloadJob.GetStatus()
+	ExpectEq(jobStatus.Name, downloader.NOT_STARTED)
+	ExpectLt(jobStatus.Offset, offset)
+	ExpectEq(n, 0)
+	ExpectNe(nil, err)
+	ExpectTrue(strings.Contains(err.Error(), util.InvalidFileDownloadJobErrMsg))
 }
 
-func (t *cacheHandleTest) TestIsSequentialWhenOffsetDiffIsEqualToMaxAllowed() {
-	t.ch.isSequential = true
-	t.ch.prevOffset = 5
-	currentOffset := 5 + downloader.ReadChunkSize
+func (cht *cacheHandleTest) Test_Read_Sequential() {
+	dst := make([]byte, ReadContentSize)
+	offset := int64(cht.object.Size - ReadContentSize)
+	cht.cacheHandle.isSequential = true
+	cht.cacheHandle.prevOffset = offset - util.MiB
+
+	// Since, it's a sequential read, hence will wait to download till requested offset.
+	_, err := cht.cacheHandle.Read(context.Background(), cht.object, true, offset, dst)
 
-	ExpectEq(true, t.ch.IsSequential(int64(currentOffset)))
+	jobStatus := cht.cacheHandle.fileDownloadJob.GetStatus()
+	ExpectGe(jobStatus.Offset, offset)
+	cht.verifyContentRead(offset, dst)
+	ExpectEq(nil, err)
 }
 
-// TODO (princer): write test which validates download flow in the cache_handle.read()
+func (cht *cacheHandleTest) Test_Read_SequentialToRandom() {
```

Comment:
please add cases
(1)  where downloadForRandom is False, read is random but there is cache hit.
(2) cache is checked if Download job is in Cancelled state.